transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,378 @@
+#endif
+
+//#define DEBUG
+
+#include <jni.h>
+#include <unistd.h>
+#include <errno.h>
+#include <libaio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include "io_netty_channel_libaio_DirectFileDescriptorController.h"
+#include "exception_helper.h"
+
+struct io_control {
+    io_context_t ioContext;
+    struct io_event * events;
+
+    // This is used to make sure we don't return IOCB while something else is using them
+    // this is to guarantee the submits could be done concurrently with polling
+    pthread_mutex_t iocbLock;
+
+    // a resuable pool of iocb
+    struct iocb ** iocb;
+    int queueSize;
+    int iocbPut;
+    int iocbGet;
+    int used;
+};
+
+jclass errorInfoClass = NULL;
+jmethodID errorInfoConstr = NULL;
+
+/**
+  This is being invoked by Native.c /  JNI_OnLoad..
+
+  There is only one point of entrance of Loading the library,
+  this will be a hook so extra stuff that needs to be loaded here.
+*/
+void directFile_JNI_OnLoad(JNIEnv* env) {
+    errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
+    if (errorInfoClass == NULL) {
+       return;
+    }
+
+    // The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    // and if the class was unloaded from GC
+    errorInfoClass = (jclass)(*env)->NewGlobalRef(env, (jobject)errorInfoClass);
+
+    errorInfoConstr = (*env)->GetMethodID(env, errorInfoClass, "<init>", "(Ljava/lang/Object;ILjava/lang/String;)V");
+
+    if (errorInfoConstr == NULL) {
+       return;
+    }
+
+    /// The ErrorInfoClass is barely used. The VM would crash in the event of an error without this GlobalRef
+    errorInfoConstr = (jmethodID)(*env)->NewGlobalRef(env, (jobject)(errorInfoConstr));
+}
+
+void directFile_JNI_OnUnLoad(JNIEnv* env) {
+    // Deleting global refs so their classes can be GCed
+    if (errorInfoConstr != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoConstr);
+    }
+
+    if (errorInfoClass != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoClass);
+    }
+}
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer) {
+    struct io_control * ioControl = (struct io_control *) (*env)->GetDirectBufferAddress(env, pointer);
+    return ioControl;
+}
+
+/**
+ remove an iocb from the pool of IOCBs. Returns null if full
+*/
+static inline struct iocb * getIOCB(struct io_control * control) {
+    struct iocb * iocb = 0;
+
+    pthread_mutex_lock(&(control->iocbLock));
+
+    #ifdef DEBUG
+       fprintf (stderr, "getIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    if (control->used < control->queueSize) {
+        control->used++;
+        iocb = control->iocb[control->iocbGet++];
+
+        if (control->iocbGet >= control->queueSize) {
+           control->iocbGet = 0;
+        }
+    } else {
```

Comment:
@Scottmitch Honestly I don't like it for Aesthetic reasons. An else inside the #ifdef... seems like the if within the else... I don't like it..  :) maybe it's my OCD kicking on here.. lol

I only needed this when developing the initial version. that case will issue a return false after my last change, I can just get rid of the entire else. you ok with that?